File: src/main/java/com/voloshko/algorithms/dynamic/LevenshteinDistance.java
package com.voloshko.algorithms.dynamic;

/**
 * Levenshtein distance: a string metric for measuring the difference between two sequences.
 * Computational complexity: O(nm)
 *
 * @author avoloshko
 */
public class LevenshteinDistance {
  static public int findDistance(String s1, String s2) {
    int[][] result = new int[s1.length() + 1][s2.length() + 1];

    for (int i = 0; i <= s1.length(); ++i) {
      result[i][0] = i;
    }
    for (int j = 0; j <= s2.length(); ++j) {
      result[0][j] = j;
    }

    for (int i = 0; i < s1.length(); ++i) {
      for (int j = 0; j < s2.length(); ++j) {
        if (s1.charAt(i) == s2.charAt(j))
          result[i + 1][j + 1] = result[i][j];
        else
          result[i + 1][j + 1] = Math.min(Math.min(result[i + 1][j], result[i][j + 1]), result[i][j]) + 1;
      }
    }

    return result[s1.length()][s2.length()];
  }
}

File: src/main/kotlin/com/voloshko/algorithms/dynamic/LevenshteinDistance.kt
package com.voloshko.algorithms.dynamic

import kotlin.math.min

enum class Weight(val value: Int) {
    ADD(1),
    DEL(1),
    REP(1)
}

fun levenshteinDistance(from: String, to: String): Int {

    val mat = Array(from.length + 1, { IntArray(to.length + 1) })

    for (i in 0..from.length) mat[i][0] = i;
    for (i in 0..to.length) mat[0][i] = i;

    for (i in 1..from.length) {
        for (j in 1..to.length) {
            mat[i][j] = if (from[i - 1] == to[j - 1]) mat[i - 1][j - 1]
            else min(mat[i - 1][j - 1]
                    + Weight.REP.value, min(mat[i][j - 1]
                    + Weight.DEL.value, mat[i - 1][j] + Weight.ADD.value))
        }
    }

    return mat[from.length][to.length];
}